### ISSUE TITLE
linter: (no-param-reassign) false-positive detection when assigning to this

### ISSUE THREAD

**Issue by dotSality** [oxc-project/oxc#22300]
### What version of Oxlint are you using?

1.63.0

### What command did you run?

oxlint ./src

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "overrides": [
    {
      "files": ["**/*.{js,ts,vue}"],
      "rules": {
        "no-param-reassign": [
          "error",
          {
            "props": true
          }
        ]
      }
    }
  ]
}
```

### What happened?

`no-param-reassign` incorrectly flags assignments to `this[key]` as a violation. ESLint correctly ignores these cases since `this` is not a parameter.

**Case 1 — Pinia store action**:
```
import { defineStore } from "pinia";

const useMockStore = defineStore("mock", {
  state: () => ({ value: null }),
  actions: {
    setProp(key: number, value: number) {
      this[key] = value; // oxc: Assignment to property of function parameter 'key'.
    },
  },
});
```

**Case 2 - Explicit this**:
a) TypeScript:
```
function Mock<T>(this: Record<PropertyKey, T>, key: PropertyKey, value: T) {
  this[key] = value; // oxc: Assignment to property of function parameter 'key'.
}
```
b) JavaScript:
```
function Mock(key, value) {
    this[key] = value; // oxc: Assignment to property of function parameter 'key'.
}
```

**Actual**: no-param-reassign error is reported on this[key] = value.
**Expected**: No error. this is not a function parameter. ESLint with the same config does not flag these cases.
